vm/src/core/mod.rs

```diff
@@ -1267,6 +1276,34 @@ impl<'a, 'e> PatternTranslator<'a, 'e> {
         self.0.allocator.arena.alloc(expr)
     }
 
+    fn compile_literal<'p>(
+        &mut self,
+        default: &'a Expr<'a>,
+        variables: &[&'a Expr<'a>],
+        equations: &[Equation<'a, 'p>],
+    ) -> &'a Expr<'a> {
+        let patterns: Vec<_> = equations
+            .iter()
+            .map(|equation| match equation.patterns.first().unwrap().value {
+                ast::Pattern::Literal(ref l) => Pattern::Literal(l.clone()),
+                ast::Pattern::Ident(ref id) => Pattern::Ident(id.clone()),
+                _ => unreachable!()
+            }).collect();
+
+        let alternatives = patterns.into_iter()
+            .zip(equations.iter().map(|e| e.result))
```

Comment:
Instead of `e.result`, this should be the same `Expr` calculated by `compile_variable` (take the first binding in that function `let expr` and make that be the right hand side of each `Alternative`).

If you are wondering why, it is because it recurses down into all the patterns we need to match on *after* the current pattern. The algorithm keeps on recursing like that (taking the first column of equations, creating a core pattern and recursing on the rest) and when `equations` is empty, then `e.result` is returned which is the expression that is on the right hand side of the AST alterantive.

Reply:
This is maybe not the problem you have right now though since you will only notice it if you have a test with nested patterns.

Reply:
So the nested pattern handling is a bit tricky still, I will try and explain as best I can but If you still have trouble I can jump in and finish that part.

Essentially you want to take this first part of `compile_constructor` https://github.com/gluon-lang/gluon/blob/0909139a2a80389a56d6546e921aef9e52abf9cc/vm/src/core/mod.rs#L1120-L1140 to group all identical `Literal` patterns together to make a `Map<Literal, Vec<Equation>>`.

The for each entry in this map (just like for the constructor case) you generate one `Alternative` with the literal as the key and use the `Vec<Equation>` value to generate the `expr` field of `Alternative` https://github.com/gluon-lang/gluon/blob/0909139a2a80389a56d6546e921aef9e52abf9cc/vm/src/core/mod.rs#L1178 . You can do more or less the same thing as in `compile_constructor` but you don't have to add the nested fields (since there are none). https://github.com/gluon-lang/gluon/blob/0909139a2a80389a56d6546e921aef9e52abf9cc/vm/src/core/mod.rs#L1162-L1165 .

Also you only need to handle `complete == false` since there is no way to have a complete match on literals.